File: docs/check-api-changes.py
#!/usr/bin/env python3
"""
Check for FreeFeed API changes against the current reference.

Clones the server, extracts routes, diffs against docs/freefeed-api.json,
and outputs a JSON change report to stdout.

Usage: python3 docs/check-api-changes.py
Output: JSON report to stdout, progress to stderr
"""

import json
import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

REPO_URL = "https://github.com/FreeFeed/freefeed-server.git"
BRANCH = "stable"

SCRIPT_DIR = Path(__file__).resolve().parent
PROJECT_DIR = SCRIPT_DIR.parent
REFERENCE = SCRIPT_DIR / "freefeed-api.json"
EXTRACT_SCRIPT = SCRIPT_DIR / "extract-routes.mjs"


def run(cmd, **kwargs):
    return subprocess.run(cmd, check=True, **kwargs)


def main():
    if not REFERENCE.exists():
        print(f"Error: {REFERENCE} not found", file=sys.stderr)
        sys.exit(1)

    work_dir = tempfile.mkdtemp()
    try:
        # Clone server
        print(f"Cloning FreeFeed server ({BRANCH} branch)...", file=sys.stderr)
        run(["git", "clone", "--branch", BRANCH, "--depth", "1", REPO_URL, work_dir],
            capture_output=True)

        # Get server rev
        result = subprocess.run(
            ["git", "rev-parse", "HEAD"], cwd=work_dir,
            capture_output=True, text=True, check=True)
        server_rev = result.stdout.strip()

        # Extract routes
        shutil.copy(EXTRACT_SCRIPT, work_dir)
        extracted_path = os.path.join(work_dir, "extracted.json")
        run(["node", "extract-routes.mjs"], cwd=work_dir,
            env={**os.environ, "SERVER_REV": server_rev, "OUT_FILE": extracted_path},
            capture_output=True)

        with open(extracted_path) as f:
            extracted = json.load(f)

        print(f"Extracted {extracted['meta']['total_endpoints']} endpoints "
              f"from server rev {server_rev[:12]}.", file=sys.stderr)
    finally:
        shutil.rmtree(work_dir, ignore_errors=True)

    # Diff against reference
    with open(REFERENCE) as f:
        current = json.load(f)

    new_eps = {f"{e['method']} {e['path']}": e for e in extracted["endpoints"]}
    cur_eps = {f"{e['method']} {e['path']}": e for e in current["endpoints"]}

    report = {
        "server_rev": extracted["meta"]["server_rev"],
        "reference_rev": current["meta"]["server_rev"],
        "added": [],
        "removed": [],
        "scope_changed": [],
        "unchanged_count": 0,
    }

    for eid, ep in sorted(new_eps.items()):
        if eid not in cur_eps:
            report["added"].append({"id": eid, "scopes": ep["scopes"]})
        elif set(ep["scopes"]) != set(cur_eps[eid].get("scopes", [])):
            report["scope_changed"].append({
                "id": eid,
                "old_scopes": cur_eps[eid].get("scopes", []),
                "new_scopes": ep["scopes"],
            })
        else:
            report["unchanged_count"] += 1

    for eid in sorted(cur_eps):
        if eid not in new_eps:
            report["removed"].append({"id": eid})

    report["total_changes"] = (
        len(report["added"]) + len(report["removed"]) + len(report["scope_changed"])
    )

    print(json.dumps(report, indent=2))

    n = report["total_changes"]
    if n == 0:
        print("No changes detected.", file=sys.stderr)
    else:
        print(f"Changes: {n} total", file=sys.stderr)
        if report["added"]:
            print(f"  +{len(report['added'])} added", file=sys.stderr)
        if report["removed"]:
            print(f"  -{len(report['removed'])} removed", file=sys.stderr)
        if report["scope_changed"]:
            print(f"  ~{len(report['scope_changed'])} scope changed", file=sys.stderr)


if __name__ == "__main__":
    main()

File: scripts/check-api-changes.mjs
#!/usr/bin/env node

/**
 * Check for FreeFeed API changes against the current reference.
 *
 * Clones the server, extracts routes, diffs against src/lib/freefeed-api.json,
 * and outputs a JSON change report to stdout.
 *
 * Usage: node scripts/check-api-changes.mjs
 * Output: JSON report to stdout, progress to stderr
 */

import { execFileSync } from 'node:child_process';
import { copyFileSync, mkdtempSync, readFileSync, rmSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import { fileURLToPath } from 'node:url';

const REPO_URL = 'https://github.com/FreeFeed/freefeed-server.git';
const BRANCH = 'stable';

const SCRIPT_DIR = fileURLToPath(new URL('.', import.meta.url));
const PROJECT_DIR = join(SCRIPT_DIR, '..');
const REFERENCE = join(PROJECT_DIR, 'src', 'lib', 'freefeed-api.json');
const EXTRACT_SCRIPT = join(SCRIPT_DIR, 'extract-routes.mjs');

function main() {
  // Verify reference file exists
  let currentRaw;
  try {
    currentRaw = readFileSync(REFERENCE, 'utf-8');
  } catch {
    process.stderr.write(`Error: ${REFERENCE} not found\n`);
    process.exit(1);
  }

  const workDir = mkdtempSync(join(tmpdir(), 'freefeed-api-'));
  let extracted;

  try {
    // Clone server
    process.stderr.write(`Cloning FreeFeed server (${BRANCH} branch)...\n`);
    execFileSync('git', ['clone', '--branch', BRANCH, '--depth', '1', REPO_URL, workDir], {
      stdio: 'pipe',
    });

    // Get server rev
    const serverRev = execFileSync('git', ['rev-parse', 'HEAD'], {
      cwd: workDir,
      encoding: 'utf-8',
      stdio: 'pipe',
    }).trim();

    // Extract routes
    copyFileSync(EXTRACT_SCRIPT, join(workDir, 'extract-routes.mjs'));
    const extractedPath = join(workDir, 'extracted.json');
    execFileSync('node', ['extract-routes.mjs'], {
      cwd: workDir,
      env: { ...process.env, SERVER_REV: serverRev, OUT_FILE: extractedPath },
      stdio: 'pipe',
    });

    extracted = JSON.parse(readFileSync(extractedPath, 'utf-8'));

    process.stderr.write(
      `Extracted ${extracted.meta.total_endpoints} endpoints ` +
        `from server rev ${serverRev.slice(0, 12)}.\n`,
    );
  } finally {
    rmSync(workDir, { recursive: true, force: true });
  }

  // Diff against reference
  const current = JSON.parse(currentRaw);

  const newEps = new Map();
  for (const e of extracted.endpoints) {
    newEps.set(`${e.method} ${e.path}`, e);
  }

  const curEps = new Map();
  for (const e of current.endpoints) {
    curEps.set(`${e.method} ${e.path}`, e);
  }

  const report = {
    server_rev: extracted.meta.server_rev,
    reference_rev: current.meta.server_rev,
    added: [],
    removed: [],
    scope_changed: [],
    unchanged_count: 0,
  };

  for (const eid of [...newEps.keys()].sort()) {
    const ep = newEps.get(eid);
    const cur = curEps.get(eid);
    if (!cur) {
      report.added.push({ id: eid, scopes: ep.scopes });
    } else if (!setsEqual(ep.scopes, cur.scopes ?? [])) {
      report.scope_changed.push({
        id: eid,
        old_scopes: cur.scopes ?? [],
        new_scopes: ep.scopes,
      });
    } else {
      report.unchanged_count += 1;
    }
  }

  for (const eid of [...curEps.keys()].sort()) {
    if (!newEps.has(eid)) {
      report.removed.push({ id: eid });
    }
  }

  report.total_changes =
    report.added.length + report.removed.length + report.scope_changed.length;

  process.stdout.write(JSON.stringify(report, null, 2) + '\n');

  const n = report.total_changes;
  if (n === 0) {
    process.stderr.write('No changes detected.\n');
  } else {
    process.stderr.write(`Changes: ${n} total\n`);
    if (report.added.length) {
      process.stderr.write(`  +${report.added.length} added\n`);
    }
    if (report.removed.length) {
      process.stderr.write(`  -${report.removed.length} removed\n`);
    }
    if (report.scope_changed.length) {
      process.stderr.write(`  ~${report.scope_changed.length} scope changed\n`);
    }
  }
}

/** Compare two arrays as sets (order and duplicates ignored) */
function setsEqual(a, b) {
  const setA = new Set(a);
  const setB = new Set(b);
  if (setA.size !== setB.size) return false;
  for (const item of setA) {
    if (!setB.has(item)) return false;
  }
  return true;
}

main();
